lib/log4js.js

```diff
@@ -87,15 +92,25 @@ function shutdown(cb) {
   // not being able to be drained because of run-away log writes.
   enabled = false;
 
-  // Call each of the shutdown functions in parallel
+  // Clone out to maintain a reference
   const appendersToCheck = Array.from(appenders.values());
+
+  // Reset immediately to prevent leaks
+  appenders.init();
+  categories.init();
```

Comment:
Is `categories.init();` even necessary, considering that `appenders.init();` sets up categories as well? No matter I still think it makes sense to call both here anyway.

Reply:
Also I never cloned the entire repo to check this logic here, but this means an stdout appender is added immediately back. Just making sure that cannot cause any trouble?

Reply:
> Is `categories.init();` even necessary, considering that `appenders.init();` sets up categories as well? No matter I still think it makes sense to call both here anyway.

I thought so too. But apparently, `appenders` and `categories` are separate. 
Both have their own cache and add their own listeners to the `configuration` object.

So there is a need to clear both to reset tot the same factory default state when you first `require(log4js)`.

## appenders
https://github.com/log4js-node/log4js-node/blob/0aed197dd36868d0a83e43f44bd9439dc85fda54/lib/appenders/index.js#L21

https://github.com/log4js-node/log4js-node/blob/0aed197dd36868d0a83e43f44bd9439dc85fda54/lib/appenders/index.js#L129

## categories
https://github.com/log4js-node/log4js-node/blob/0aed197dd36868d0a83e43f44bd9439dc85fda54/lib/categories.js#L6

https://github.com/log4js-node/log4js-node/blob/0aed197dd36868d0a83e43f44bd9439dc85fda54/lib/categories.js#L165

## configuration
https://github.com/log4js-node/log4js-node/blob/0aed197dd36868d0a83e43f44bd9439dc85fda54/lib/configuration.js#L17-L20
https://github.com/log4js-node/log4js-node/blob/0aed197dd36868d0a83e43f44bd9439dc85fda54/lib/configuration.js#L37-L48

Reply:
> Also I never cloned the entire repo to check this logic here, but this means an stdout appender is added immediately back. Just making sure that cannot cause any trouble?

Yup, I've checked. In fact, this is the same factory default behavior when you first `require('log4js')`.

https://github.com/log4js-node/log4js-node/blob/0aed197dd36868d0a83e43f44bd9439dc85fda54/lib/appenders/index.js#L105

So what I did was to wrap around it so that I can call `appenders.init()` in `log4js.shutdown()`.
Hence, calling `log4js.shutdown()` will now reset back to same factory default state when you first `require('log4js')`.

```diff
- setup({ appenders: { out: { type: 'stdout' } }, categories: { default: { appenders: ['out'], level: 'trace' } } });
+ const init = () => {
+   setup({ appenders: { out: { type: 'stdout' } }, categories: { default: { appenders: ['out'], level: 'trace' } } });
+ };
+ init();
...
+ module.exports.init = init;
```